namespace_validate.go

```diff
@@ -0,0 +1,118 @@
+package kubeauth
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	k8s_yaml "k8s.io/apimachinery/pkg/util/yaml"
+)
+
+// This exists so we can use a mock namespace validation when running tests
+type namespaceValidator interface {
+	ValidateLabels(context.Context, *http.Client, string, string) (bool, error)
+}
+
+type namespaceValidateFactory func(*kubeConfig) namespaceValidator
+
+// This is the real implementation that calls the kubernetes API
+type namespaceValidateAPI struct {
+	config *kubeConfig
+}
+
+func namespaceValidateAPIFactory(config *kubeConfig) namespaceValidator {
+	return &namespaceValidateAPI{
+		config: config,
+	}
+}
+
+func (v *namespaceValidateAPI) ValidateLabels(ctx context.Context, client *http.Client, namespace string, namespaceSelector string) (bool, error) {
+	labelSelector, err := makeLabelSelector(namespaceSelector)
```

Comment:
It doesn't look like we are guarding against empty namespace, nor empty namespaceSelector. I think we might want to return an error if either one is empty.

Reply:
The guarding step happens in path_role.go where `bound_service_account_namespaces`  and `bound_service_account_namespace_selector` are set. With `bound_service_account_namespace_selector`, the  PR allows either but not both of these fields to be empty.
https://github.com/hashicorp/vault-plugin-auth-kubernetes/blob/ca554dd48b7c13bc95f8610ba4ba590765aa368e/path_role.go#L318-L321

`ValidateLabels(...)` is used in path_login.go where the request's token namespace meets the namespace condition. I think we don't need the guarding step here as  path_login.go only references those fields. https://github.com/hashicorp/vault-plugin-auth-kubernetes/blob/ca554dd48b7c13bc95f8610ba4ba590765aa368e/path_login.go#L342-L356

Let me know if I misunderstood your point anywhere or how I can improve the guarding step.